### PR TITLE
Add basic design for the API

### PR DESCRIPTION
**Pull request by ctyar**
<div dir="rtl">
منتقل شده از:

https://github.com/softwaretalks/pangul/issues/6#issuecomment-461735093

این PR رو من از طرف @MCSH ایجاد می‌کنم صرفا برای اینکه بتونیم راجع به تصمیمات گرفته شده در ارتباط با API بیشتر صحبت کنیم

</div>

### CODE REVIEW

File: swagger.txt [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+tags:
+- name: url
+  description: Checks url and everything related to them
+- name: report
+  description: Reports 
+schemes:
+- "https"
+- "http"
+paths:
+  /url:
```
Comment (psycho-ir) [inline swagger.txt]: how about plural naming for resources?
i.e
`/urls`
`/urls/{url}`
`...`
we can follow this guideline for the API design.
https://opensource.zalando.com/restful-api-guidelines/

File: swagger.txt [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    get:
+      tags:
+      - url
+      summary: Get all of URLS
+      description: ""
+      operationId: listURL
+      produces:
+      - application/json
+      responses:
+        201:
```
Comment (psycho-ir) [inline swagger.txt]:
```suggestion
        200:
```

File: swagger.txt [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+          description: Created
```
Comment (psycho-ir) [inline swagger.txt]:
```suggestion
          description: List of all eligible urls
```

File: swagger.txt [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/URL'
+        409:
```
Comment (psycho-ir) [inline swagger.txt]: and `400` if the URL is malformed or any other thing in request is invalid.

File: swagger.txt [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+          description: URL exists
+    post:
+      tags:
+      - url
+      summary: Add a new url to pangul
+      description: ""
+      operationId: addURL
+      consumes:
```
Comment (psycho-ir) [inline swagger.txt]: as `application/json` is the default content-type we can introduce once before the `paths` and just override it if it was needed

File: swagger.txt [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        200:
+          description: Ok
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Report'
+        404:
+          description: Not found
+  /report:
+    post:
```
Comment (psycho-ir) [inline swagger.txt]: Is this part of the public contract?
Do we want to let the users to use this API?

File: swagger.txt [hunk 7]
```diff
@@ -0,0 +1,182 @@
+swagger: "2.0"
+info:
+  description: "Pnagul API schema."
```
Comment (ctyar) [inline swagger.txt]: Typo:
```suggestion
  description: "Pangul API schema."
```

File: swagger.txt [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+  version: "1.0.0"
+  title: "Pnagul"
```
Comment (ctyar) [inline swagger.txt]:
```suggestion
  title: "Pangul"
```

File: swagger.txt [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+  description: Reports 
+schemes:
+- "https"
+- "http"
+paths:
+  /url:
+    get:
+      tags:
+      - url
+      summary: Get all of URLS
```
Comment (ctyar) [inline swagger.txt]:
```suggestion
      summary: Returns all URLs
```

File: swagger.txt [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+      produces:
+      - application/json
+      responses:
+        200:
+          description: Ok
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Report'
+  /report/{id}:
```
Comment (ctyar) [inline swagger.txt]: I don't think there is any value in displaying a single report. Everywhere we are going to aggregate all the (recent) reports of a URL. I think we should just get the URL here and return all reports associated with it and remove the `/url/{id}/reports`.

File: swagger.txt (hunk 4 above)
Comment (ctyar) [reply swagger.txt]: We don't need this error code, we are just returning all URLs here.

File: swagger.txt [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+      operationId: addURL
+      consumes:
+      - application/json
+      produces:
+      - application/json
+      parameters:
+      - in: body
+        name: body
+        required: true
+        schema:
```
Comment (ctyar) [inline swagger.txt]: The `schema` here should be different, we don't the `id` when creating a new `URL`

File: swagger.txt [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+          $ref: '#/definitions/URL'
+      responses:
+        201:
+          description: Created
+          schema:
+            $ref: '#/definitions/URL'
+        409:
```
Comment (psycho-ir) [inline swagger.txt]: and `400` if the URL is malformed or any other thing in request is invalid.

File: swagger.txt (hunk 10 above)
Comment (psycho-ir) [reply swagger.txt]: And probably for `/urls/{id}/reports` we need to return data aggregated.
if we do return all the reports/pings then the list size for just 1 day will be 1440 items. 
Some thoughts:
- return only failed pings (downside is that in the worse case can have the same size (1440 items per day)
- return ranges of failures and successes e.g. between 10:30 and 11:30 pings failed but before and after that everything was good. that means for this specific example we have only 3 items in the list for one day

**Pull request merged**